### ISSUE TITLE
Error when trying to set backend as OpenCL

### ISSUE THREAD

**Issue by dhaverporth** [arrayfire/arrayfire#2482]
I am trying to ascertain if an algorithm will run faster on multiple CPUs vs. a GPU.  I coded the algorithm in ArrayFire because I can switch between modes easily.  The Algorithm ran fine on GPU but when trying to use OpenCL for the multiple cores I am getting an error I don't understand.  I am using the Python wrappers.  The error occurs when I make the call `af.set_backend('opencl')`.  

The error is as follows:

> File "usr/local/lib/python3.6/site-packages/arrayfire/library.py", line 604, in set_backend
> safe_call(backend.get().af_set_backend(backend.get_id(name)))
>File "usr/local/lib/python3.6/site-packages/arrayfire/util.py", line 79, in safe_call
> raise RuntimeError(to_str(err_str))
>RuntimeError: Error in af_err unified::AFSymbolManager::setBackend(af::Backend)
>In file src/api/unified/symbol_manager.cpp:

The funny thing is that the symbol_manager.cpp file is not on my machine so I don't understand how it is generating an error.

I also have the following enviornment variables set:

> AF_OPENCL_DEVICE_TYPE=CPU
>AF_DISABLE_GRAPHICS=1
>AF_PATH=/usr/local/bin/arrayfire/include

**Comment by 9prady9** [comment]
It seems like you are trying to use CPU via OpenCL backend. If that is correct, you will need OpenCL drivers provided by your CPU(Intel/AMD). Can please you post the whole code. If it is larger project, perhaps github project link.

**Comment by dhaverporth** [comment]
Unfortunately I can't post the code since it is sensitive but we are running on a machine with 80 cores and 1 Tesla v100 GPU.  This is a linux system.  We are trying to run all 80 cores using OpenCL via arrayfire.  We think we have the intel drivers installed.  In the /etc/OpenCL/vendors/ folder we place a intel64.icd file.  The icd file contains the link to /opt/intel/opencl-1.2-6.4.0.25/libintelocl.so.  Is libintelocl.so the intel driver?

**Comment by 9prady9** [comment]
What does the `af::info()` output look like ? How many Intel OpenCL devices are showing up in `af::info` output ? Python wrapper call is [arrayfire.device.info](http://arrayfire.org/arrayfire-python/arrayfire.device.html?highlight=info#arrayfire.device.info)

**Comment by dhaverporth** [comment]
Sorry for the delayed reply.   For some reason arrayfire will no longer import in python.  I will try reinstalling and see if that would fix the problem.

**Comment by dhaverporth** [comment]
Ok sorry for the delay.  One of the things I tried to fix the problem caused things to go really bad.  Here is the af.device.info:

ArrayFire v3.6.1 (CUDA, 64-bit Linux, build 4fcd0dd)
Platform: CUDA Toolkit 9.2, Driver: 396.44
[0] Tesla v100-PCIE-16GB, 16161 MB, CUDA Compute 7.0

**Comment by 9prady9** [comment]
@dhaverporth Sorry about the delay, I was out of town. If what you pasted above is the entirety of af::info output, then only NVIDIA GPU is the only OpenCL device visible to ArrayFire at the moment. Have you tried using any CPU(opencl) using any other standalone OpenCL application. If not, you can try it to list the available devices using tools like [clinfo](https://github.com/Oblomov/clinfo)

Most of the times when device isn't listed by ArrayFire, it means the library isn't available via PATH for ArrayFire to load.

Note: You can install opencl runtime for CPUs from [here](https://software.intel.com/en-us/articles/opencl-drivers#cpu-section)

**Comment by dhaverporth** [comment]
9prady9:  I've added /etc/OpenCL/vendors to the PATH and LD_LIBRARY_PATH and it hasn't fixed the issue.  I am new to OpenCL so maybe that is not the right path to add.

**Comment by dhaverporth** [comment]
I also did install clinfo and ran it.  It found the intel drivers and listed the correct amount of cores.

**Comment by 9prady9** [comment]
@dhaverporth Can you please share the entire clinfo output.

`/etc/OpenCL/vendors` should not be added manually by user, it is standard location where vendor icds are looked for.

Given below is an example from my machine which has an APU and NVIDIA GPU.
command: `locate libOpenCL.so`
```
/opt/cuda/targets/x86_64-linux/lib/libOpenCL.so
/opt/cuda/targets/x86_64-linux/lib/libOpenCL.so.1
/opt/cuda/targets/x86_64-linux/lib/libOpenCL.so.1.1
/usr/lib/libOpenCL.so
/usr/lib/libOpenCL.so.1
...
```
command: `ls /etc/OpenCL/vendors/`
```
amdocl64.icd  amdocl-orca64.icd  nvidia.icd
```
As you have already noticed, the icd's contain the opencl driver indirections. On my system, both the redirections point to libs under `/usr/lib`.

From one of you earlier comment, the intel opencl driver seems to be located under ` /opt/intel/opencl-1.2-6.4.0.25/` which doesn't seem to be located in your PATH environment variable. Or at least, there should be a symlink under `/usr/lib` or `/usr/lib64` to this file under `/opt`.  Do you see such symlink ? That is the problem I think. 

But interestingly, you claim that clinfo is reporting Intel GPU as well. Hence, I requested you to share the entire clinfo output so that I can have a look at it.

**Comment by dhaverporth** [comment]
@9prady9 :  I did try adding /opt/intel/opencl-1.2-6.4.0.25/lib64/ to the PATH enviornment variable and it didn't fix the problem.  In that directory the libOpenCL.so is structured as follows:

libOpenCL.so is soft linked to libOpenCL.so.1 in the same folder
libOpenCL.so.1 is soft linked to libOpenCL.so.2.0 in the same folder
libOpenCL.so.2.0 is the actual file listed in that directory

The machine that I am working on is a secure machine which is not connected to the internet unfortunately so I cannot copy and paste the clinfo output but I will list some of the output here:

> Platform Name:  NVIDIA CUDA
>Platform Vendor: NVIDIA Corporation
>Platform Version: OpenCL1.2 CUDA 9.0.176
>Platform Profile: FULL_PROFILE
>
>Platform Name: Intel (R) OpenCL
>Platform vendor: Intel (R) Corporation
>Platform version OpenCL 1.2 LINUX
>Platform Profile:  FULL_PROFILE
>...
>NULL Platform behavior
>clGetPlatformInfo(NULL, CL_PLATFORM_NAME,...) NVIDIA CUDA
>clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...) Success [NV]
>clCreateContext(NULL, ...) [default]    Success [NV]
>clCreateContext(Null,..) [other]  <checkNullCtx: 2308: Create context with device from non-default platform:  error -2>
>clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU) No devices found in platform
>clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU) No Platform
>clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR) no Device Found
>clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM) No Device Found
>clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL) No platform
>
>ICD loader properties
>ICD loader Name   openCL ICD Loader
>ICD loader Vendor  OCL Icd free software
>ICD loader Version  2.2.12
>ICD loader Profile OpenCL 2.2
> NOTE: your OpenCL library declares to support OpenCL 2.2, but is seems to support up to OpenCL 2.1 Only

**Comment by 9prady9** [comment]
You can try the following. Use `ldd <path  to clinfo>` to see what libraries it is using and do the same with arrayfire exectuable.

Sample output of ldd for clinfo on my machine is given below.
```
ldd /usr/bin/clinfo
	linux-vdso.so.1 (0x00007ffe547e3000)
	libOpenCL.so.1 => /opt/cuda/lib64/libOpenCL.so.1 (0x00007f2743f10000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f2743eb0000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f2743ce8000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f2743cc0000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f2744338000)

```

**Comment by dhaverporth** [comment]
Here is the output for `ldd /usr/bin/clinfo`:

> linux-vsdo.so.1
>libOpenCL.so.1 => /usr/local/bin/arrayfire/lib/libOpenCL.so.1
>libdl.so.2 => /lib64/libdl.so.2
>libc.so.6 => /lib64/libc.so.6

**Comment by 9prady9** [comment]
@dhaverporth Have you been able to use any of our C++ examples that ship with our binary ? with OpenCL backend of course.

**Comment by dhaverporth** [comment]
@9prady9 :  There is something definitely wrong with the OpenCL.  Looking over the clinfo again it looks like when it attempts to do a clCreateContext command for the CPU it is creating an error.  The error code it was giving is "-2".  I found a website that said what "-2" means and it means that the CL device is not available.  I talked with out system admin to make sure he downloaded the right OpenCL drivers.   It turns out he simply used the drivers from an older machine with the drivers dated from 2016.  However the Intel Xeon Gold 6138 processors that we are using were introduced in 2017 so it is possible the driver he installed is too old.  Since our computers are on secure networks and not on the Internet there is a delay getting new software to the machines so I probably won't know until tomorrow if the new drivers fixes the problem.  Since moving software up to the secure network is a pain our sys admins typically cannibalize software from other machines instead of downloading the latest software from the internet.

**Comment by dhaverporth** [comment]
@9prady9 :  So we have installed the new drivers and the specific error that I mentioned in the beginning of the thread is no longer there.  clinfo is also not showing the "-2" error and the clCreateContext is working.  However, Arrayfire will no longer import in Python, the process just hangs and doesn't do anything.  Before we could import ArrayFire and we only had the error when we tried to running `af.set_backend('opencl')`.  So it seems like we fixed one problem and created another.  I will go ahead and close this particular issue but I may create another one for the issue we are currently seeing.  Thanks for all you help!
